File: keybot/reddit.py
#!/usr/bin/env python3

import sys, os
import http.server
import concurrent.futures
import webbrowser
import urllib.parse
import uuid
import logging
import datetime
import typing

import requests
import requests.auth

def getLogger(name=__name__, level=logging.DEBUG):
    logger = logging.getLogger(name)
    logger.setLevel(level)
    handler = logging.StreamHandler(sys.stderr)
    formatter = logging.Formatter("%(levelname)s - %(message)s")
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    return logger

Logger = getLogger()

class AuthCallbackHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        self.server.Params = urllib.parse.parse_qs(
            urllib.parse.urlparse(self.path).query)
        self.send_response(200)
        self.end_headers()
        self.wfile.write("It's a trap!".encode("utf-8"))
        self.wfile.flush()
        self.close_connection = True

    def log_message(self, format, *args, **kargs):
        pass

class RedditQuerier(object):
    UrlBase = "https://oauth.reddit.com"
    UserAgent = "desktop:org.darksair.keybot:0.0.1 (by /u/darksair)"

    def __init__(self):
        self.Token = None
        self.RefreshToken = None
        self.HttpSession = requests.Session()
        self.HttpSession.headers["User-Agent"] = self.UserAgent

    def authenticateUserLess(self, client_id, client_secret):
        Url = "https://www.reddit.com/api/v1/access_token"
        Res = self.HttpSession.post(
            Url, data={"grant_type": "client_credentials"},
            auth=requests.auth.HTTPBasicAuth(client_id, client_secret))
        try:
            Res.raise_for_status()
        except requests.HTTPError as Err:
            raise RuntimeError("Failed to get token: " + str(Err))

        Data = Res.json()
        self.Token = Data["access_token"]
        self.HttpSession.headers["Authorization"] =  "bearer " + self.Token

    def authenticate(self, client_id, client_secret):
        StateStr = uuid.uuid1().hex
        InitPayload = {"client_id": client_id,
                       "response_type": "code",
                       "state": StateStr,
                       "redirect_uri": "http://localhost:31416/",
                       "duration": "temporary",
                       "scope": "identity edit flair history mysubreddits "
                       "privatemessages read report save submit subscribe vote "
                       "wikiedit wikiread"}

        InitUrl = "https://www.reddit.com/api/v1/authorize"

        PermUrl = InitUrl + '?' + urllib.parse.urlencode(InitPayload)
        webbrowser.open(PermUrl)
        CallbackServer = http.server.HTTPServer(("", 31416), AuthCallbackHandler)
        try:
            CallbackServer.handle_request()
        except KeyboardInterrupt:
            CallbackServer.socket.close()

        if hasattr(CallbackServer, "Params"):
            if CallbackServer.Params["state"][0] != StateStr:
                raise RuntimeError("Failed to get token. State doesn't match.")
            if "error" in CallbackServer.Params:
                raise RuntimeError("Failed to get code: " +
                                   CallbackServer.Params["error"][0])

            Code = CallbackServer.Params["code"][0]

        TokenPayload = {"grant_type": "authorization_code",
                        "code": Code,
                        "redirect_uri": "http://localhost:31416/"}

        Res = self.HttpSession.post(
            "https://www.reddit.com/api/v1/access_token", data=TokenPayload,
            auth=requests.auth.HTTPBasicAuth(client_id, client_secret))
        try:
            Res.raise_for_status()
        except requests.HTTPError as Err:
            raise RuntimeError("Failed to get token: " + str(Err))
        Data = Res.json()
        self.Token = Data["access_token"]
        self.RefreshToken = Data.get("refresh_token")
        self.HttpSession.headers.update({"Authorization": "bearer " + self.Token,
                                         "User-Agent": self.UserAgent})

    def logout(self, client_id, client_secret):
        Payload = {"token": self.Token, "token_type_hint": "access_token"}
        Res = requests.post("https://www.reddit.com/api/v1/revoke_token",
                            data=Payload, headers={"User-Agent": self.UserAgent},
                            auth=requests.auth.HTTPBasicAuth(client_id, client_secret))
        try:
            Res.raise_for_status()
        except requests.HTTPError as Err:
            raise RuntimeError("Failed to logout: " + str(Err))

    def _urlPreprocess(self, url):
        Component = urllib.parse.urlparse(url)
        if Component.query == "":
            Url = url + "?raw_json=1"
        else:
            Url = url + "&raw_json=1"

        if not url.startswith(self.UrlBase):
            Url = self.UrlBase + Url

        return Url

    def get(self, url, *args, **kargs):
        Res = self.HttpSession.get(self._urlPreprocess(url), *args, **kargs)
        try:
            Res.raise_for_status()
        except requests.HTTPError as Err:
            raise RuntimeError("Failed to query {}: {}".format(url, Err))
        else:
            return Res

    def post(self, url, *args, **kargs):
        Res = self.HttpSession.post(self._urlPreprocess(Url), *args, **kargs)
        try:
            Res.raise_for_status()
        except requests.HTTPError as Err:
            raise RuntimeError("Failed to query {}: {}".format(url, Err))
        else:
            return Res

RedditQuery = RedditQuerier()

class RedditPost(object):
    def __init__(self):
        self.Title = ""
        self.Text = ""
        self.Author = ""
        self.Score = 0
        self.Url = ""           # URI to the post itself
        self.Link = ""          # Link in the post
        self.HideScore = False
        self.ID = ""            # “Fullname”
        self.CountComments = 0
        self.TimeCreate = 0     # type: datetime.datetime
        self.Sub = ""

    @property
    def FullName(self) -> str:
        return self.ID

    @property
    def IsLink(self) -> bool:
        if self.Link.startswith(
                "https://www.reddit.com/r/{}/comments".format(self.Sub)):
            if self.Link[len("https://www.reddit.com"):] == self.Url:
                return False
        return True

    @property
    def ShortUrl(self) -> str:
        UID = self.ID.split('_', 1)[1]
        return "https://reddit.com/r/{}/comments/{}/".format(self.Sub, UID)

class PostSorting(object):
    Hot = "hot"
    New = "new"

class SubredditInfo(object):
    def __init__(self, name):
        self.Name = name

    @property
    def UrlName(self):
        return "/r/" + self.Name

    def list(self, sorting: str, before: str=None, after: str=None, limit=50) \
        -> typing.List[RedditPost]:
        Query = {"g": "GLOBAL"}
        if before is not None:
            Query["before"] = before
        elif after is not None:
            Query["after"] = after

        Res = RedditQuery.get("{}/{}.json".format(self.UrlName, sorting),
                              params=Query)
        Data = Res.json()
        Posts = []
        for Item in Data["data"]["children"]:
            Data = Item["data"]
            Post = RedditPost()
            Post.Title = Data["title"]
            Post.Text = Data["selftext"]
            Post.Author = Data["author"]
            Post.Score = Data["score"]
            Post.Url = Data["permalink"]
            Post.Link = Data["url"]
            Post.HideScore = Data["hide_score"]
            Post.ID = Data["name"]
            Post.CountComments = Data["num_comments"]
            Post.TimeCreate = datetime.datetime.fromtimestamp(Data["created_utc"])
            Post.Sub = Data["subreddit"]
            Posts.append(Post)

        return Posts

File: src/reddit.rs
use std::vec::Vec;
use std::collections::HashMap;
use std::time;
use std::fmt;

// use log::info;
use reqwest;
use reqwest::blocking as requests;
use serde_json;
use uuid;
use chrono::prelude::*;

use crate::error::Error;
use crate::simple_http_server;

#[allow(dead_code)]
pub struct RedditQuerier
{
    token: String,
    refresh_token: Option<String>,
    client_id: String,
    client_secret: String,
}

impl RedditQuerier
{
    const URL_BASE: &'static str = "https://oauth.reddit.com";
    const USER_AGENT: &'static str = "desktop:org.darksair.keybot:0.0.1 (by /u/darksair)";

    pub async fn fromUserlessAuthentication(
        client_id: &str, client_secret: &str) -> Result<Self, Error>
    {
        let url = "https://www.reddit.com/api/v1/access_token";
        let payload = [("grant_type", "client_credentials"),];
        let client = reqwest::Client::new();
        let res = match client.post(url).header("User-Agent", Self::USER_AGENT)
            .form(&payload).basic_auth(client_id, Some(client_secret))
            .send().await
        {
            Err(_) => { return Err(error!(
                RedditError, "Failed to authenticate userless (post)")); },
            Ok(res) => res,
        };

        if let Err(e) = res.error_for_status_ref()
        {
            return Err(error!(
                RedditError, format!("Failed to authenticate userless ({})",
                                     e.status().unwrap().as_u16())));
        }

        let data: HashMap<String, serde_json::Value> = res.json().await.map_err(|e| {
            error!(RedditError, format!("Failed to authenticate userless: {}", e))
        })?;

        Ok(Self {
            token: String::from(data["access_token"].as_str().unwrap()),
            refresh_token: None,
            client_id: client_id.to_owned(),
            client_secret: client_secret.to_owned(),
        })
    }

    #[allow(dead_code)]
    pub fn fromAuthentication(client_id: &str, client_secret: &str)
                              -> Result<Self, Error>
    {
        let now = time::SystemTime::now().duration_since(time::UNIX_EPOCH)
            .map_err(|_| {error!(RedditError, "failed to get time")})?;
        let state_str: String = uuid::Uuid::new_v1(
            uuid::v1::Timestamp::from_unix(
                uuid::v1::Context::new(42), now.as_secs(), now.subsec_nanos()),
            &[1, 2, 3, 4, 5, 6]).map_err(
            |_| { error!(RedditError, "failed to generate UUID") })?
            .to_string();

        let payload_init = [
            ("client_id", client_id),
            ("response_type", "code"),
            ("state", &state_str),
            ("redirect_uri", "http://localhost:31416/"),
            ("duration", "temporary"),
            ("scope", "identity edit flair history mysubreddits \
privatemessages read report save submit subscribe vote \
wikiedit wikiread")];

        let perm_url = reqwest::Url::parse_with_params(
            "https://www.reddit.com/api/v1/authorize", &payload_init).unwrap();
        println!("Please open the following URI in your browser:\n\n{}",
                 perm_url);

        let server = simple_http_server::SimpleHttpServer::new(31416);
        let params = server.handleOne()?;

        if params.contains_key("error")
        {
            return Err(error!(RedditError, format!(
                "Failed to authenticate: {}", params["error"])));
        }

        if params["state"] != state_str
        {
            return Err(error!(RedditError, "Invalid state string"));
        }

        let code = &params["code"];

        let payload_token = [
            ("grant_type", "authorization_code"),
            ("code", &code),
            ("redirect_uri", "http://localhost:31416/")];

        let client = requests::Client::new();
        let res = client.post("https://www.reddit.com/api/v1/access_token")
            .header("User-Agent", Self::USER_AGENT)
            .form(&payload_token).basic_auth(client_id, Some(client_secret))
            .send().map_err(|_| {
                error!(RedditError, "Failed to post token request")})?;

        res.error_for_status_ref().map_err(|e| {
            error!(RedditError, format!("Failed to get token: {}", e))})?;

        let data: HashMap<String, serde_json::Value> = res.json().map_err(|e| {
            error!(RedditError, format!("Failed to parse token response: {}", e))
        })?;

        Ok(Self {
            token: String::from(data["access_token"].as_str().unwrap()),
            refresh_token: Some(String::from(
                data["refresh_token"].as_str().unwrap())),
            client_id: client_id.to_owned(),
            client_secret: client_secret.to_owned(),
        })
    }

    pub fn urlPreprocess(url_raw: &str) -> Result<String, Error>
    {
        let url = reqwest::Url::parse(&("http://localhost".to_owned() + url_raw))
            .map_err(|_| error!(
                RedditError, format!("Failed to parse url {}", url_raw)))?;
        let mut result: String;
        if url.query().is_none()
        {
            result = url_raw.to_owned() + "?raw_json=1";
        }
        else
        {
            result = url_raw.to_owned() + "&raw_json=1";
        }

        if !url_raw.starts_with(Self::URL_BASE)
        {
            result = Self::URL_BASE.to_owned() + &result;
        }
        Ok(result)
    }

    pub async fn query(&self, req_builder: reqwest::RequestBuilder)
                       -> Result<reqwest::Response, Error>
    {
        let mut result = req_builder.header("User-Agent", Self::USER_AGENT);
        result = result.header(
            "Authorization",
            "bearer ".to_owned() + self.token.as_ref());
        let res = result.send().await.map_err(
            |e| {error!(RedditError,
                        format!("Failed to send request: {}", e))})?;
        res.error_for_status().map_err(
            |e| {error!(RedditError, format!("Query failed: {}", e))})
    }

    pub async fn logout(self) -> Result<(), Error>
    {
        let payload = [("token", self.token.as_ref()),
                       ("token_type_hint", "access_token")];
        let client = reqwest::Client::new();
        let res = client.post("https://www.reddit.com/api/v1/revoke_token")
            .header("User-Agent", Self::USER_AGENT)
            .form(&payload).basic_auth(&self.client_id, Some(&self.client_secret))
            .send().await.map_err(|_| {
                error!(RedditError, "Failed to post logout request")})?;

        res.error_for_status().map_err(|e| {
            error!(RedditError, format!("Failed to logout: {}", e))}).map(|_| ())
    }
}

pub struct Post
{
    pub title: String,
    pub text: String,
    pub author: String,
    pub score: i32,
    pub url: String,            // URI to the post itself
    pub link: String,           // Link in the post
    pub hide_score: bool,
    pub id: String,
    pub count_comments: u32,
    pub time_create: DateTime<Utc>,
    pub sub: String,
}

impl Post
{
    #[allow(dead_code)]
    pub fn new() -> Self
    {
        Self {
            title: String::new(),
            text: String::new(),
            author: String::new(),
            score: 0,
            url: String::new(),
            link: String::new(),
            hide_score: false,
            id: String::new(),
            count_comments: 0,
            time_create: Utc.timestamp(0, 0),
            sub: String::new(),
        }
    }

    pub fn fullName(&self) -> &str
    {
        &self.id
    }

    pub fn isLink(&self) -> bool
    {
        if self.link.starts_with(
            &format!("https://www.reddit.com/r/{}/comments", self.sub))
        {
            if self.link["https://www.reddit.com".len()..] == self.url
            {
                return false;
            }
        }
        true
    }

    pub fn shortUrl(&self) -> String
    {
        let uid = self.id.splitn(1, "_").last().unwrap();
        format!("https://reddit.com/r/{}/comments/{}/", self.sub, uid)
    }
}

impl fmt::Debug for Post
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result
    {
        write!(f, "Post{{{}}}", self.url)
    }
}

makeIntEnum!
{
    PostSorting {hot, new,} with u8,
    derive(Copy, Clone, PartialEq)
}

pub struct Subreddit
{
    name: String
}

impl Subreddit
{
    pub fn new(sub: &str) -> Self
    {
        Self {name: sub.to_owned()}
    }

    fn urlName(&self) -> String
    {
        "/r/".to_owned() + &self.name
    }

    pub async fn list(&self, querier: &RedditQuerier, sorting: PostSorting,
                      before: Option<&str>, after: Option<&str>)
                      -> Result<Vec<Post>, Error>
    {
        let mut query = vec![("g", "GLOBAL"),];
        if let Some(before_id) = before
        {
            query.push(("before", before_id));
        }
        if let Some(after_id) = after
        {
            query.push(("after", after_id));
        }

        let client = reqwest::Client::new();
        let res = querier.query(
            client.get(&RedditQuerier::urlPreprocess(
                &format!("{}/{}.json", self.urlName(), sorting))?)
                .query(&query)).await?;
        let data: serde_json::Value = res.json().await.map_err(
            |_| {error!(RedditError, "Invalid JSON in listing")})?;
        // println!("{}", serde_json::to_string_pretty(&data).unwrap());
        Ok(data["data"]["children"].as_array().unwrap().iter().map(
            |post_data_wrapper|
            {
                let data = &post_data_wrapper["data"];
                Post {
                    title: data["title"].as_str().unwrap().to_owned(),
                    text: data["selftext"].as_str().unwrap().to_owned(),
                    author: data["author"].as_str().unwrap().to_owned(),
                    score: data["score"].as_i64().unwrap() as i32,
                    url: data["permalink"].as_str().unwrap().to_owned(),
                    link: data["url"].as_str().unwrap().to_owned(),
                    hide_score: data["hide_score"].as_bool().unwrap(),
                    id: data["name"].as_str().unwrap().to_owned(),
                    count_comments: data["num_comments"].as_u64().unwrap() as u32,
                    time_create: Utc.timestamp(
                        data["created_utc"].as_f64().unwrap() as i64, 0),
                    sub: data["subreddit"].as_str().unwrap().to_owned(),
                }
            }).collect())
    }
}
